### PR TITLE
Reasoner integration

### PR DESCRIPTION
**Pull request by jmsfltchr**
## What is the goal of this PR?

Provide the final integration between components (along with the necessary fixes), such that the reasoner runs end-to-end for basic reasoning queries.

## What are the changes implemented in this PR?

- Implement real traversals, removing the last of the mocking in reasoner, and thereby running the first working reasoner queries successfully! This required functionality to implement bound traversals - for use when supplying a set of known ids to add to a traversal. Critical since we execute conjunctions knowing a partial answer in all of reasoner's resolvers.
- Various bug fixes necessary to get the reasoner up and running:
  - Update `iterationInferredAnswer` so that reiteration terminates.
  - AnswerState merging of answers was duplicated and done incorrectly. Now `aggregate` methods are removed, in favour of `withInitial()`.
  - Removed some overly-aggressive assertions.
  - Behaviour of RuleResolver and RootResolver when they received exhausted messages.
  - Add return statements to while loops so that they terminate.
  - Missing variable from materialisation when materialising rules.
  - Unification of inequality value constraints.
- Add basic tests of reasoner queries, `ReasonerTest`.

### CODE REVIEW

**Review by lolski — APPROVED** [review]

File: pattern/Conjunction.java [hunk 1]
```diff
@@ -123,6 +123,12 @@ public Traversal traversal(List<Identifier.Variable.Name> filter) {
         return traversal;
     }
 
+    public Traversal traversal() {
+        Traversal traversal = new Traversal();
+        variableSet.forEach(variable -> variable.addTo(traversal));
+        return traversal;
+    }
```
Comment (flyingsilverfin) [inline pattern/Conjunction.java:130]: how is this different from the above method with an empty set of filters? if you want both, then one should call the other and add the filters inside

File: logic/resolvable/Concludable.java [hunk 2]
```diff
@@ -249,6 +249,12 @@ else if (a.getType().isDouble())
         return predicateFn;
     }
 
+    protected void addValueEqualsRequirements(Unifier.Builder unifierBuilder, Set<ValueConstraint<?>> values) {
+        for (ValueConstraint<?> value : equalsConstraints(values)) {
```
Comment (flyingsilverfin) [inline logic/resolvable/Concludable.java:253]: maybe able to inline `equalsConstraints` if not used anywhere else

File: reasoner/Reasoner.java [hunk 3]
```diff
@@ -82,7 +82,7 @@ private boolean isInfer() {
                                                 Context.Query context) {
         ResourceIterator<Conjunction> conjunctions = iterate(disjunction.conjunctions());
         if (!context.options().parallel()) return conjunctions.flatMap(conj -> iterator(conj, filter));
-        else return produce(conjunctions.flatMap(conj -> producers(conj, filter)).toList());
+        else return produce(conjunctions.flatMap(conj -> producers(conj, filter)).toList()).distinct();
```
Comment (flyingsilverfin) [inline reasoner/Reasoner.java:85]: we wanted to remove this no?

File: reasoner/resolution/answer/AnswerState.java [hunk 4]
```diff
@@ -58,6 +58,13 @@ public int hashCode() {
         return Objects.hash(conceptMap);
     }
 
+    @Override
+    public String toString() {
+        return "AnswerState{" +
+                "conceptMap=" + conceptMap +
+                '}';
+    }
```
Comment (flyingsilverfin) [inline reasoner/resolution/answer/AnswerState.java:66]: can actually be deleted, impl by each subclass

File: test/integration/reasoner/ReasonerTest.java [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        grakn.databases().create(database);
+    }
+
+    @After
+    public void tearDown() {
+        grakn.close();
+    }
+
+    @Test
+    public void test_basic() {
```
Comment (flyingsilverfin) [inline test/integration/reasoner/ReasonerTest.java:71]: basic is a bad name imo, `test_no_rules()` better?

File: test/integration/reasoner/ReasonerTest.java [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                    assertEquals("person", a.get("p1").asThing().getType().getLabel().scopedName());
+                    assertEquals("person", a.get("p2").asThing().getType().getLabel().scopedName());
+                    assertEquals("name", a.get("na").asAttribute().getType().getLabel().scopedName());
+                });
+
+                assertEquals(2, ans.size());
+            }
+        }
+    }
+}
```
Comment (flyingsilverfin) [inline test/integration/reasoner/ReasonerTest.java:184]: eheheheehehehehehehehehehehe

**Review by flyingsilverfin — CHANGES_REQUESTED** [review]

File: reasoner/Reasoner.java (hunk 3 above)
Comment (jmsfltchr) [reply reasoner/Reasoner.java:85]: Yeah but then the basic test fails! So I put it back.

File: pattern/Conjunction.java (hunk 1 above)
Comment (jmsfltchr) [reply pattern/Conjunction.java:130]: Fair point!

File: logic/resolvable/Concludable.java (hunk 2 above)
Comment (jmsfltchr) [reply logic/resolvable/Concludable.java:253]: It's used in 3 places which is why I pulled it out: in Has, Value and Attribute.

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**